### PR TITLE
incr.comp.: Make DepNode `Copy` and valid across compilation sessions

### PR DESCRIPTION
**Pull request by michaelwoerister**
This PR moves `DepNode` to a representation that does not need retracing and thus simplifies comparing dep-graphs from different compilation sessions. The code also gets a lot simpler in many places, since we don't need the generic parameter on `DepNode` anymore.  See https://github.com/rust-lang/rust/issues/42294 for details.

~~NOTE: Only the last commit of this is new, the rest is already reviewed in https://github.com/rust-lang/rust/pull/42504.~~

This PR is almost done but there are some things I still want to do:
- [x] Add some module-level documentation to `dep_node.rs`, explaining especially what the `define_dep_nodes!()` macro is about.
- [x] Do another pass over the dep-graph loading logic. I suspect that we can get rid of building the `edges` map and also use arrays instead of hash maps in some places.

cc @rust-lang/compiler 
r? @nikomatsakis

### CODE REVIEW

**Comment by Mark-Simulacrum** [comment]
Looks like tidy failed here:

```
[00:03:21] tidy error: /checkout/src/librustc_incremental/persist/preds/mod.rs:69: line longer than 100 chars

```

**Comment by michaelwoerister** [comment]
OK, this should be ready for review.

File: src/librustc/dep_graph/dep_node.rs [hunk 1]
```diff
@@ -17,6 +17,58 @@ use rustc_data_structures::stable_hasher::{StableHasher, HashStable};
 use ich::StableHashingContext;
 use std::hash::Hash;
 
+//! This module defines the `DepNode` type which the compiler uses to represent
```
Comment (nikomatsakis) [inline src/librustc/dep_graph/dep_node.rs]: :clinking_glasses: love the comments

File: src/librustc/dep_graph/dep_node.rs [hunk 2]
```diff
@@ -8,79 +8,280 @@
...
 
-macro_rules! try_opt {
-    ($e:expr) => (
-        match $e {
-            Some(r) => r,
-            None => return None,
+// erase!() just makes tokens go away. It's used to specify which macro argument
+// is repeated (i.e. which sub-expression of the macro we are in) but don't need
+// to actually use any of the arguments.
+macro_rules! erase {
```
Comment (nikomatsakis) [inline src/librustc/dep_graph/dep_node.rs]: cute.

**Comment by nikomatsakis** [comment]
This is really nice. r=me once errors are resolved.

**Pull request merged**